### PR TITLE
http: introduce addEncodedMetadata()

### PR DESCRIPTION
**Pull request by soya3129**
Description: introduce addEncodedMetadata(). Newly added metadata by addEncodedMetadata() will only be passed to filters following the filter where the metadata are added.
Risk Level: low.
Testing: integration tests.
Docs Changes: Updated metadata docs.
Release Notes: Introduce addEncodeMetadata() function.

### CODE REVIEW

**Comment by soya3129** [comment]
cc @birenroy

File: source/common/http/conn_manager_impl.h [hunk 1]
```diff
@@ -173,8 +173,8 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     // either because [de|en]codeHeaders() of the current filter returns StopAllIteration or because
     // [de|en]codeHeaders() adds new metadata to [de|en]code, but we don't know
     // [de|en]codeHeaders()'s return value yet. The storage is created on demand.
-    std::unique_ptr<MetadataMapVector> saved_request_metadata_;
-    std::unique_ptr<MetadataMapVector> saved_response_metadata_;
+    std::unique_ptr<MetadataMapVector> saved_request_metadata_ = nullptr;
+    std::unique_ptr<MetadataMapVector> saved_response_metadata_ = nullptr;
```
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.h]: Is this necessary?  If so let's use the {nullptr} syntax you use for request_medata_map_vector_

File: source/common/http/http1/codec_impl.cc [hunk 2]
```diff
@@ -164,6 +164,10 @@ void StreamEncoderImpl::encodeData(Buffer::Instance& data, bool end_stream) {
 
 void StreamEncoderImpl::encodeTrailers(const HeaderMap&) { endEncode(); }
 
+void StreamEncoderImpl::encodeMetadata(const MetadataMapVector&) {
+  ENVOY_LOG_MISC(error, "HTTP1 doesn't support metadata");
```
Comment (alyssawilk) [inline source/common/http/http1/codec_impl.cc]: So if an H2 upstream sends medatadata to an H1 data source, we'll log an error message for every frame?  I think we want to handle that in a way that's less painful.  I think ideally we'd error log once (or better yet with pow-2 backoff) and increment a stat, but I don't think Envoy has backoff logging yet so maybe just a stat for now?

File: test/integration/filters/response_metadata_filter.cc [hunk 3]
```diff
@@ -16,42 +16,39 @@ class ResponseMetadataStreamFilter : public Http::PassThroughFilter {
 public:
   // Inserts one new metadata_map.
   Http::FilterHeadersStatus encodeHeaders(Http::HeaderMap&, bool) override {
-    Http::MetadataMap metadata_map = {
-        {"headers", "headers"}, {"duplicate", "duplicate"}, {"remove", "remove"}};
+    Http::MetadataMap metadata_map = {{"headers", "headers"}, {"duplicate", "duplicate"}};
     Http::MetadataMapPtr metadata_map_ptr = std::make_unique<Http::MetadataMap>(metadata_map);
-    decoder_callbacks_->encodeMetadata(std::move(metadata_map_ptr));
```
Comment (alyssawilk) [inline test/integration/filters/response_metadata_filter.cc]: Are there HCM unit tests we should update as well?

File: source/common/http/conn_manager_impl.h (hunk 1 above)
Comment (soya3129) [reply source/common/http/conn_manager_impl.h]: Fixed. Thanks!

File: source/common/http/http1/codec_impl.cc (hunk 2 above)
Comment (soya3129) [reply source/common/http/http1/codec_impl.cc]: Good idea! Fixed in a separate [PR](https://github.com/envoyproxy/envoy/pull/7801).

File: test/integration/filters/response_metadata_filter.cc (hunk 3 above)
Comment (soya3129) [reply test/integration/filters/response_metadata_filter.cc]: I think adding new metadata was only covered in integration test.

File: source/common/http/http1/codec_impl.cc (hunk 2 above)
Comment (alyssawilk) [reply source/common/http/http1/codec_impl.cc]: Cool.  If you want to change this to a no-op I'd be fine with it landing as-is, though it should probably get a Matt Klein pass as well.

Comment (soya3129) [reply source/common/http/http1/codec_impl.cc]: Maybe I can sync to master after #7801 is merged?

Comment (mattklein123) [reply source/common/http/http1/codec_impl.cc]: Yeah agreed this should be a NOP + stat, so fine to review the other change first and then come back to this one.

**Comment by soya3129** [comment]
Merged h1 stats change. PTAL. Thanks!

**Pull request merged**